File: code/education/course/src/api/student-hub.ts
import axios from 'axios';

/** 学生中心读接口在后端不可达时要尽快失败，页面才能进入课程快照状态。 */
const READ_TIMEOUT_MS = 8000;

export interface StudentMessage {
  id: string;
  title: string;
  body: string;
  category: string;
  is_read: boolean;
  link?: string | null;
  created_at: string;
}

export interface StudyGroupItem {
  id: string;
  name: string;
  description: string;
  member_count: number;
  course_name?: string | null;
  my_role: string;
  updated_at: string;
}

export interface PracticeTopicSummary {
  subject: string;
  topic: string;
  sessions: number;
  total_questions: number;
  correct_count: number;
  avg_score: number;
  last_practiced_at?: string | null;
}

export interface PracticeSummary {
  total_sessions: number;
  total_questions: number;
  correct_rate: number;
  subjects: string[];
  topics: PracticeTopicSummary[];
  assignment_completed: number;
  assignment_total: number;
}

export interface AchievementItem {
  id: string;
  code: string;
  title: string;
  description: string;
  icon: string;
  points_awarded: number;
  earned_at: string;
}

export interface AchievementPayload {
  total_points: number;
  level: number;
  next_level_points: number;
  data: AchievementItem[];
  count: number;
}

function unwrap<T>(res: unknown): T {
  const payload = res as { data?: T };
  return (payload.data ?? res) as T;
}

export async function fetchStudentMessages(limit = 20) {
  const res = await axios.get('/api/student-hub/messages', {
    params: { limit },
    timeout: READ_TIMEOUT_MS,
  });
  return unwrap<{ data: StudentMessage[] }>(res).data ?? [];
}

export async function fetchStudyGroups() {
  const res = await axios.get('/api/student-hub/groups', { timeout: READ_TIMEOUT_MS });
  return unwrap<{ data: StudyGroupItem[] }>(res).data ?? [];
}

export async function fetchPracticeSummary() {
  const res = await axios.get('/api/student-hub/practice/summary', {
    timeout: READ_TIMEOUT_MS,
  });
  return unwrap<PracticeSummary>(res);
}

export async function fetchAchievements() {
  const res = await axios.get('/api/student-hub/achievements', {
    timeout: READ_TIMEOUT_MS,
  });
  return unwrap<AchievementPayload>(res);
}

File: code/backend/app/schemas/student_hub.py
from __future__ import annotations

from pydantic import BaseModel, Field


class StudentNotificationPublic(BaseModel):
    id: str
    title: str
    body: str
    category: str
    is_read: bool
    link: str | None = None
    created_at: str


class StudentNotificationsPublic(BaseModel):
    data: list[StudentNotificationPublic]
    count: int
    unread_count: int


class StudyGroupMemberPublic(BaseModel):
    student_id: str
    student_name: str
    role: str


class StudyGroupPublic(BaseModel):
    id: str
    name: str
    description: str
    tc_id: str | None = None
    course_name: str | None = None
    member_count: int
    my_role: str
    members: list[StudyGroupMemberPublic] = Field(default_factory=list)
    updated_at: str


class StudyGroupsPublic(BaseModel):
    data: list[StudyGroupPublic]
    count: int


class PracticeTopicSummary(BaseModel):
    subject: str
    topic: str
    sessions: int
    total_questions: int
    correct_count: int
    avg_score: float
    last_practiced_at: str | None = None


class PracticeSummaryPublic(BaseModel):
    total_sessions: int
    total_questions: int
    correct_rate: float
    subjects: list[str] = Field(default_factory=list)
    topics: list[PracticeTopicSummary] = Field(default_factory=list)
    assignment_completed: int
    assignment_total: int


class AchievementPublic(BaseModel):
    id: str
    code: str
    title: str
    description: str
    icon: str
    points_awarded: int
    earned_at: str


class AchievementsPublic(BaseModel):
    total_points: int
    level: int
    next_level_points: int
    data: list[AchievementPublic]
    count: int
